File: plugins/oracle/scripts/mcp-fleet/services/slack.mjs
// Slack workspace probe.
//
// Slack's OAuth path is closed for non-Marketplace MCP apps, so we use
// the cookie-extraction approach that korotovsky/slack-mcp-server expects:
// a `d` cookie (xoxd-prefixed) plus an `xoxc-` token harvested from the
// browser session. The user logs into a single workspace inside this
// fresh Chrome profile; we then walk every cookie and pull what's needed.
//
// References:
//   https://github.com/korotovsky/slack-mcp-server -- env vars consumed:
//     SLACK_MCP_XOXC_TOKEN, SLACK_MCP_XOXD_TOKEN

import { launchProfile } from "../lib/playwright-launcher.mjs";
import { ask, askSecret, confirm } from "../lib/prompt.mjs";
import { upsert } from "../lib/workspaces-store.mjs";

export async function probe({ label }) {
  const wsLabel = label || (await ask("Slack workspace label (e.g. 'acme-prod'): "));
  if (!wsLabel) throw new Error("workspace label is required");

  process.stdout.write(
    `\nLaunching Chrome with an isolated profile for slack/${wsLabel}.\n` +
    `Sign in to the ONE Slack workspace you want to bind to this label.\n` +
    `Once you see the workspace's main view, return here and press Enter.\n\n`,
  );

  const { context, page, close } = await launchProfile({
    service: "slack",
    workspace: wsLabel,
    startUrl: "https://slack.com/signin",
  });

  await ask("Press Enter once you are signed in and the workspace is loaded ");

  const cookies = await context.cookies();
  await close();

  const dCookie = cookies.find((c) => c.name === "d" && c.domain.endsWith("slack.com"));
  if (!dCookie) {
    process.stderr.write("Could not find Slack `d` cookie. Sign-in may have failed.\n");
    const manualXoxd = await askSecret("Paste xoxd token (input hidden): ");
    const manualXoxc = await askSecret("Paste xoxc token (input hidden): ");
    return persist(wsLabel, { xoxd: manualXoxd, xoxc: manualXoxc });
  }

  // The `d` cookie value IS the xoxd token (Slack set-cookies it pre-prefixed).
  // The xoxc token is workspace-keyed and lives in localStorage under the
  // localConfig_v2 key, indexed by the team ID parsed from the URL path.
  // Source: korotovsky/slack-mcp-server docs/01-authentication-setup.md
  // (verified 2026-05-13). Earlier `window.boot_data.api_token` paths are
  // stale.
  let xoxc = "";
  try {
    const { page: page2, close: close2 } = await launchProfile({
      service: "slack",
      workspace: wsLabel,
      startUrl: "https://app.slack.com/client",
    });
    // Wait for the URL to actually load a team -- the path becomes
    // /client/<TEAM_ID>/... once Slack picks the workspace.
    await page2.waitForURL(/\/client\/[A-Z0-9]+/, { timeout: 30000 }).catch(() => {});
    xoxc = await page2.evaluate(() => {
      try {
        // eslint-disable-next-line no-undef
        const cfg = JSON.parse(localStorage.localConfig_v2);
        // eslint-disable-next-line no-undef
        const m = document.location.pathname.match(/^\/client\/([A-Z0-9]+)/);
        if (!m) return "";
        const team = cfg.teams && cfg.teams[m[1]];
        return (team && team.token) || "";
      } catch { return ""; }
    }).catch(() => "");
    await close2();
  } catch { /* fall through to manual paste */ }

  if (!xoxc) {
    process.stderr.write(
      "Could not auto-extract xoxc token from localStorage. " +
      "Open https://app.slack.com/client in the same profile, open DevTools, " +
      "and run:\n" +
      "  JSON.parse(localStorage.localConfig_v2).teams[" +
      "document.location.pathname.match(/^\\/client\\/([A-Z0-9]+)/)[1]].token\n" +
      "Paste the resulting xoxc-... token here.\n",
    );
    xoxc = await askSecret("Paste xoxc token (input hidden): ");
  }

  return persist(wsLabel, { xoxd: dCookie.value, xoxc });
}

function persist(label, { xoxd, xoxc }) {
  if (!xoxd || !xoxc) throw new Error("missing xoxd or xoxc token");
  return upsert({
    service: "slack",
    label,
    kind: "slack-cookie",
    credentials: {
      SLACK_MCP_XOXD_TOKEN: xoxd,
      SLACK_MCP_XOXC_TOKEN: xoxc,
    },
    profileDir: "", // populated by upsert via service+label lookup
  });
}

// Allow `node services/slack.mjs <label>` directly for debugging.
if (import.meta.url === `file://${process.argv[1]}`) {
  probe({ label: process.argv[2] }).then(
    (e) => { process.stdout.write(`Persisted slack workspace ${e.label}\n`); process.exit(0); },
    (e) => { process.stderr.write(`${e.message}\n`); process.exit(1); },
  );
}

File: plugins/oracle/scripts/mcp-fleet-py/mcp_fleet/services/slack.py
"""Slack workspace probe.

Slack's OAuth path is closed for non-Marketplace MCP apps. We use the
cookie-extraction approach that korotovsky/slack-mcp-server expects:
a `d` cookie (the xoxd token) plus an `xoxc-` token harvested from
localStorage after navigating to app.slack.com/client.

Flow:
  1. Open chromium in an isolated profile at slack.com/signin.
  2. TUI awaits user signal "signed_in".
  3. Read cookies from context, find `d`.
  4. Navigate to app.slack.com/client, wait for the team-keyed URL,
     read localStorage.localConfig_v2.teams[<id>].token.
  5. If either auto-extraction fails, prompt user to paste manually.
  6. Validate via slack.com/api/auth.test, persist if it returns ok=true.
"""

from __future__ import annotations

import re

import httpx
from playwright.async_api import async_playwright

from .. import store
from ..paths import profile_dir
from .base import ProbeContext, ProbeError


TEAM_URL_RE = re.compile(r"/client/([A-Z0-9]+)")


async def _extract_xoxc(context, ctx: ProbeContext) -> str:
    """Open app.slack.com/client in the same context and pull xoxc from localStorage."""
    page = await context.new_page()
    try:
        await page.goto("https://app.slack.com/client", wait_until="domcontentloaded", timeout=30000)
    except Exception:
        ctx.log("app.slack.com/client did not load; xoxc auto-extract may fail")

    try:
        await page.wait_for_url(TEAM_URL_RE, timeout=30000)
    except Exception:
        ctx.log("did not see /client/<TEAM_ID> URL within 30s; trying anyway")

    xoxc: str = await page.evaluate(
        """
        () => {
          try {
            const cfg = JSON.parse(localStorage.localConfig_v2);
            const m = document.location.pathname.match(/^\\/client\\/([A-Z0-9]+)/);
            if (!m) return "";
            const team = cfg.teams && cfg.teams[m[1]];
            return (team && team.token) || "";
          } catch { return ""; }
        }
        """
    )
    try:
        await page.close()
    except Exception:
        pass
    return xoxc or ""


async def _validate(xoxc: str, xoxd: str) -> str:
    """Slack auth.test returns the team name when the token is good."""
    async with httpx.AsyncClient(timeout=15) as client:
        r = await client.post(
            "https://slack.com/api/auth.test",
            data={"token": xoxc},
            cookies={"d": xoxd},
        )
    body = r.json() if r.headers.get("content-type", "").startswith("application/json") else {}
    if not body.get("ok"):
        raise ProbeError(f"Slack auth.test failed: {body.get('error') or 'unknown'}")
    return body.get("team") or "Slack workspace"


async def probe(label: str, ctx: ProbeContext) -> store.Workspace:
    if not label:
        raise ProbeError("workspace label is required")

    user_data = profile_dir("slack", label)

    ctx.log(f"opening Chromium for slack/{label}")
    ctx.log("sign in to ONE workspace, then click 'Signed in' in the TUI")

    async with async_playwright() as p:
        context = await p.chromium.launch_persistent_context(
            str(user_data),
            headless=False,
            viewport={"width": 1280, "height": 900},
            ignore_default_args=["--enable-automation"],
            args=[
                "--no-first-run",
                "--no-default-browser-check",
                "--disable-blink-features=AutomationControlled",
            ],
        )
        page = context.pages[0] if context.pages else await context.new_page()
        try:
            await page.goto("https://slack.com/signin", wait_until="domcontentloaded", timeout=30000)
        except Exception:
            pass

        await ctx.wait_signal("signed_in")

        cookies = await context.cookies()
        d_cookie = next(
            (c for c in cookies if c.get("name") == "d" and c.get("domain", "").endswith("slack.com")),
            None,
        )

        xoxd = d_cookie.get("value") if d_cookie else ""
        if xoxd:
            ctx.log("found `d` cookie (xoxd token)")
        else:
            ctx.log("`d` cookie not found in this profile")

        xoxc = ""
        if xoxd:
            ctx.log("walking app.slack.com/client to harvest xoxc from localStorage ...")
            try:
                xoxc = await _extract_xoxc(context, ctx)
            except Exception as e:
                ctx.log(f"auto-extraction error: {e}")

        try:
            await context.close()
        except Exception:
            pass

    if not xoxd:
        ctx.log("falling back to manual xoxd paste")
        xoxd = (await ctx.paste_secret("Paste xoxd token (the `d` cookie value)")).strip()
    if not xoxc:
        ctx.log("falling back to manual xoxc paste -- in DevTools on app.slack.com/client:")
        ctx.log("  JSON.parse(localStorage.localConfig_v2).teams[")
        ctx.log("    document.location.pathname.match(/^\\/client\\/([A-Z0-9]+)/)[1]")
        ctx.log("  ].token")
        xoxc = (await ctx.paste_secret("Paste xoxc token")).strip()

    if not xoxd or not xoxc:
        raise ProbeError("missing xoxd or xoxc token after manual fallback")

    ctx.log("validating against slack.com/api/auth.test ...")
    team_name = await _validate(xoxc, xoxd)
    ctx.log(f"validated -- team: {team_name}")

    return store.upsert(store.Workspace(
        service="slack",
        label=label,
        kind="slack-cookie",
        credentials={
            "SLACK_MCP_XOXC_TOKEN": xoxc,
            "SLACK_MCP_XOXD_TOKEN": xoxd,
        },
        profileDir=str(user_data),
        discoveredName=team_name,
    ))
